**Important** This is a copy of [CENTIPEDE](git@github.com:bakerwm/CENTIPEDE.git)


## Branch

+ `master` - The original scripts 
+ `fixInf` - Fix **Inf errors** for some samples     
+ `dev` - Fix **Inf errors** and create **ggplot2 plot**    


## Why this repo?

Record the changes I have made, in order to pass the programe.

I failed to go through the [CENTIPEDE tutorial](https://github.com/slowkow/CENTIPEDE.tutorial). Finally, After a few modification for function `fitCentipede()`, I skipped the errors. 

Download the CENTIPEDE [source code](http://download.r-forge.r-project.org/src/contrib/CENTIPEDE_1.2.tar.gz).

**License**: GNU General Public License (GPL) (original) [CENTIPEDE on R forge](https://r-forge.r-project.org/projects/centipede/)


## How I ran the program?

PWM for CTCF was downlaoded from [Motif database](http://meme-suite.org/doc/download.html) on MEME website.

```
# MEME version 5.0.5

fimo --text --parse-genomic-coord CTCF.meme peak.fa | gzip > CTCF.fimo.txt.gz

cen <- CENTIPEDE.tutorial::centipede_data(
  bam_file   = "align.bam",
  fimo_file  = "CTCF.fimo.txt.gz",
  pvalue     = 1e-4,
  flank_size = 100
)

fit <- CENTIPEDE::fitCentipede(
  Xlist = list(DNase = cen$mat),
  Y     = as.matrix(data.frame(Intercept = rep(1, nrow(cen$mat))))
)

```

**Error**

```
...
Initialization of the parameters:
Error in quantile.default(x, c(TrimP, 1 - TrimP)) :
missing values and NaN's not allowed if 'na.rm' is FALSE
```

I failed to create the `fit` object for my ATAC-Seq data using the `CENTIPEDE` package. Here are the commands and errors:


## Why failed?

**0** values in `LambdaParList` and `LambdaParNoMixList` were converted to **-Inf** after `log()`. These **-Inf** will cause the script stop at [matrix + `-Inf`].

```R
LambdaParNoMixList <- lapply(Xlist, compLambdaNoMix, DampLambda=DampLambda)

LambdaParList <- lapply(Xlist, compLambda, Ez = Ez, DampLambda = DampLambda)
```

Here are the two functions:

```R
compLambdaNoMix <- function(
  X, DampLambda = 0.0
  ){
  S <- dim(X)[2]
  if(S > 1){
    ##    Ez <- Ez*LambPi+0.5*(1-LambPi) ## Dampening experiment
    Lambda <- colSums(X)
    Lambda <- Lambda/sum(Lambda)
    Lambda <- Lambda * (1 - DampLambda) + (DampLambda) * 1/S
  }else{
    Lambda <- 1
  }
  Lambda
}


compLambda <- function(
  X, Ez, DampLambda = 0.0
  ){
  S <- dim(X)[2]
  if(S > 1){
    ## Ez <- Ez*(1-DampLambda)+0.5*(DampLambda) ## Dampening experiment
    Lambda <- t(t(Ez) %*% X)
    Lambda[Lambda < 1] <- 1 # give 1 to 0 columns
    Lambda <- Lambda/sum(Lambda)
    Lambda <- Lambda * (1 - DampLambda) + (DampLambda) * 1/S
  }else{
    Lambda <- 1
  }
  Lambda
}
```

**Failed**, The **0** values after `Lambda <- colSums(X)`, and **-Inf** values generated by `log()` calulation in two functions `compLogLik1()` and `compMultiNomLogRatio()`, and the program will stopped within the two functions.

**Wired**, It means the read count on some positions was zero?

## How to fix?

Assign the minimum value of `Lambda` to the `-Inf` in functions: `compLogLik1()` and `compMultiNomLogRatio()`.

**compLogLik1()**

```
## MultiNomLogLik <-  X %*% log(Lambda) ## Multinomial for 1 and 0

## fix
fn        <- log(Lambda)
fn_noInf  <- fn[is.finite(fn)]
fn[is.infinite(fn)] <- min(fn_noInf) # min value
MultiNomLogLik      <-  X %*% fn
```

**compMultiNomLogRatio()**

```
## MultiNomLogRatio <-  X %*% log(Lambda * S) ## Multinomial for 1 and 0

## fix
fn        <- log(Lambda * S)
fn_noInf  <- fn[is.finite(fn)]
# assign median to Inf values
fn[is.infinite(fn)] <- min(fn_noInf) # min value
MultiNomLogRatio    <-  X %*% fn
```

## Finally

From the documentation, I did not found the content about `LambdaParList` or `LambdaParNoMixList`.

**Important**, We should go back to the script produce the `Xlist`. 

